---
title: "IMG/M metagenome protein exploration"
author: "Connor Morgan-Lang"
date: "April 13, 2019"
output: html_document
---

```{r setup, include=FALSE}
library(ggplot2)
library(RColorBrewer)
library(tidyr)
library(dplyr)
library(stringr)
library(knitr)
library(purrr)

Rank <- c("Root", "Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
Depth <- c(1, 2, 3, 4, 5, 6, 7, 8)

rank_depth_map <- data.frame(Rank, Depth)

MoreColours <- function(n_o, palette="BrBG") {
  ramp_func <- colorRampPalette(brewer.pal(9, palette))
  ramp_func(n_o)
}

test_str <- "Root; Archaea; Euryarchaeota; Methanomicrobia; Methanomicrobiales; Methanomicrobiaceae"
resolved_rank <- function(lineage_str) {
  taxa <- str_split(lineage_str, pattern="; ")
  len <- length(unlist(taxa))
  if (len == 0) {
    return("Unclassified")
  }
  rank <- as.character(with(rank_depth_map, Rank[Depth %in% len])) %>% 
    unlist()
  return(rank)
}
resolved_rank(test_str)

```

```{r load_data, include=FALSE}
dat <- read.table(file="~/Bioinformatics/Hallam_projects/TreeSAPP_manuscript/IMG_metagenome_TreeSAPP_output_McrABG/final_outputs/extra_annotated_marker_contig_map.tsv",
                  header=TRUE,
                  sep="\t", quote="") %>% 
  separate(col = Query, into=c("taxon_oid", "contig_name"), sep = ':') %>% 
  mutate(taxon_oid = gsub(taxon_oid, pattern = ".a", replacement = ''))

# Specific.Ecosystem intentionally not included due to most entries being either 'Unclassified'/NA or redundant
metadata <- read.table(file="~/Bioinformatics/Hallam_projects/TreeSAPP_manuscript/dereplicated_metagenomes_10_01_2017.tab",
                       header=T,
                       sep="\t", quote="", fill = T, na.strings = c('', "NA")) %>% 
  select(taxon_oid, GOLD.Analysis.Project.ID, GOLD.Sequencing.Project.ID, Altitude, Ecosystem, Ecosystem.Category, Ecosystem.Subtype, Ecosystem.Type, Geographic.Location, Isolation, Isolation.Country, Latitude, Longitude, Depth, Habitat, pH) %>% 
  mutate(Ecosystem = replace_na(as.character(Ecosystem), "Unavailable")) %>% 
  mutate(Ecosystem.Category = replace_na(as.character(Ecosystem.Category), "Unavailable")) %>% 
  mutate(Ecosystem.Subtype = replace_na(as.character(Ecosystem.Subtype), "Unavailable"))
  
mcra_dat <- merge(dat, metadata, by="taxon_oid") %>%
  filter("McrA" == Marker) %>% 
  mutate(Resolved = as.character(lapply(Confident_Taxonomy, resolved_rank))) %>%
  separate(col = Confident_Taxonomy, into = c('r', 'k', 'p', 'c', 'o', 'f', 'g', 's'), sep = "; ", fill = "right", remove = T) %>% 
  select(-Sample, -iNode, -Abundance, -LWR, -Distances, -Taxonomy, -r)
mcra_dat$Ecosystem.Subtype <- gsub("Unclassified", "Unavailable", mcra_dat$Ecosystem.Subtype)
rm(dat, metadata)

```

First off, I want to see the spread of classifications across different taxonomic ranks. Were most sequences going to Root? Each of the following variables represent the number of sequences that were classified at a rank, and the following rank was not assigned (i.e. non-cumulative).

```{r classification-depth, fig.height=2, fig.width=6, fig.align="center"}
known_df <- mcra_dat %>% 
  group_by(Resolved) %>% 
  count() %>% 
  ungroup() %>% 
  merge(rank_depth_map, by.x="Resolved", by.y="Rank") %>% 
  mutate(Proportion = n/sum(n)) %>% 
  mutate(Resolved = reorder(Resolved, Depth))

k_r <- filter(known_df, Resolved == "Root")$n
k_k <- filter(known_df, Resolved == "Kingdom")$n
k_p <- filter(known_df, Resolved == "Phylum")$n

ggplot(known_df, aes(fill=Resolved, y=Proportion, x=1)) +
  geom_bar(stat = "identity", colour="black") +
  ylab("Assigned Sequences") +
  scale_fill_brewer(palette = "PuOr") +
  coord_flip() +
  theme(panel.background = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.grid.major.x = element_blank(),
        panel.grid.minor = element_blank(),
        axis.title.y = element_blank(),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank())

```
Great! We were actually able to assign a reasonable portion of McrA sequences to a Species and over 50% of the sequences were assigned to Genus or greater resolution. I envision this single bar being a component of a multi-part figure summarizing the IMG metagenome classification results.

Where are the really novel McrA sequences (note: not divergent as in the butane- or ethane-oxidizing Mcr) from and how many are there?

```{r phylum-less, fig.align="center"}
unknowns <- mcra_dat %>% 
  filter(is.na(p)) %>% 
  group_by(Ecosystem.Category, Ecosystem.Subtype) %>% 
  count() %>% 
  ungroup() %>% 
  mutate(Percent = 100*n/sum(n)) %>% 
  mutate(Ecosystem.Subtype = reorder(Ecosystem.Subtype, Percent, FUN = sum)) %>% 
  mutate(Ecosystem.Category = reorder(Ecosystem.Category, Percent, FUN = sum))  # different functions change the colour order

n_unknown <- sum(unknowns$n)

ggplot(unknowns, aes(x=Ecosystem.Subtype, y=n, fill=Ecosystem.Category)) +
  geom_bar(stat = "identity", colour="black", size=0.5) +
  scale_fill_manual(values = MoreColours(length(unique(unknowns$Ecosystem.Category)))) +
  guides(fill=guide_legend(title="Ecosystem"),
         size=FALSE) +
  ylab("Query sequences") +
  xlab("Ecosystem Sub-category") +
  theme(panel.background = element_blank(),
        panel.grid.major.y = element_line(colour = "#bdbdbd", linetype = "dotted"),
        panel.grid.major.x = element_blank(),
        panel.grid.minor = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1))

bug_mcra <- filter(mcra_dat, Ecosystem.Category == "Arthropoda") %>%
  filter(Ecosystem.Subtype == "Gut") %>%
  filter(is.na(p)) %>%
  select(taxon_oid) %>%
  unique() %>%
  nrow()
```

There are `r n_unknown` McrA sequences that could not be classified at the rank of Phylum. Most of these are derived from Aquative environments, predominantly Hydrothermal vent, ocean and lake ecosystems. Also, a considerable number (71 or 12%) of these unknown sequences originate from `r bug_mcra` arthropod gut metagenomes.

Now onto looking at who we found.

```{r census, fig.height=5, fig.width=8, fig.align="center"}

c_census_dat <- mcra_dat %>% 
  filter(!is.na(c)) %>% 
  group_by(c) %>% 
  count() %>% 
  ungroup() %>% 
  mutate(c = reorder(c, n))
  
ggplot(c_census_dat, aes(x=c, y=n)) +
  geom_bar(stat="identity") +
  xlab("Class") +
  ylab("Count") +
  theme(panel.background = element_blank(),
        panel.grid.major.y = element_line(colour = "#bdbdbd", linetype = "dotted"),
        panel.grid.major.x = element_blank(),
        panel.grid.minor = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1))
```
The number of McrA sequences identified from each taxonomic class across all metagenome proteins in IMG/M. The vast majority belong to Methanomicrobia and Methanobacteria though there is a significant signal from some newly described lineages such as the _Hadesarchaea_, _Candidatus Methanoliparia_ and _Candidatus Methanomethylia_. Sequences that could not be classified to the rank of Class (`r sum(k_r, k_k, k_p)`) are not accounted for in this plot.

Who is inhabiting what environments? Sequences that were not classified at the rank of Order have been removed.

```{r endemism, fig.height=9, fig.width=10, fig.align="center"}

o_census_dat <- mcra_dat %>% 
  filter(!is.na(o)) %>% 
  group_by(c, o, Ecosystem.Category, Ecosystem.Subtype) %>% 
  count() %>% 
  ungroup() %>% 
  filter(n >= 2)

ggplot(o_census_dat, aes(x=Ecosystem.Subtype, y=o)) +
  geom_point(aes(size=n, fill=c),
             position = position_jitterdodge(dodge.width = 0, jitter.width = 0.25, jitter.height = 0.25),
             colour="black", pch=21) +
  scale_radius(range = c(1,10),
               breaks = c(min(o_census_dat$n),
                          round(mean(o_census_dat$n)),
                          max(o_census_dat$n))) +
  scale_fill_manual(values = MoreColours(length(unique(o_census_dat$c)))) +
  guides(fill=guide_legend(title="Class"),
         size=guide_legend(title="Count")) +
  ylab("Order") +
  xlab("Ecosystem") +
  theme(panel.background = element_blank(),
        panel.grid.major.y = element_line(colour = "#bdbdbd", linetype = "dotted"),
        panel.grid.major.x = element_line(colour = "#bdbdbd", linetype = "dotted"),
        panel.grid.minor = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1))

```

Metabolism analysis follows. What is the distribution of metabolisms?

```{r metabolical, fig.height=8, fig.width=10, fig.align="center"}

metabolic_census <- mcra_dat %>% 
  filter(!is.na(c)) %>% 
  group_by(Metabolism, c) %>% 
  count() %>% 
  ungroup() %>%
  mutate(Metabolism = reorder(Metabolism, n, FUN = sum))

ggplot(metabolic_census, aes(x=Metabolism, y=n)) +
  geom_bar(stat = "identity", colour="black", aes(fill=c)) +
  scale_fill_manual(values = MoreColours(length(unique(c_census_dat$c)))) +
  guides(fill=guide_legend(title="Class")) +
  ylab("Count") +
  xlab("Metabolism") +
  theme(panel.background = element_blank(),
        panel.grid.major.y = element_line(colour = "#bdbdbd", linetype = "dotted"),
        panel.grid.major.x = element_blank(),
        panel.grid.minor = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1))

metabolic_env <- mcra_dat %>% 
  filter(Ecosystem.Subtype != "Unclassified") %>% 
  filter(!is.na(f)) %>% 
  group_by(Metabolism, f, Ecosystem.Subtype) %>% 
  count() %>% 
  ungroup() %>% 
  filter(n >= 2) %>% 
  mutate(Ecosystem.Subtype = reorder(Ecosystem.Subtype, n))

ggplot(metabolic_env, aes(x=Ecosystem.Subtype, y=f)) +
  geom_point(aes(size=n, fill=Metabolism),
             position = position_jitterdodge(dodge.width = 0, jitter.width = 0.25, jitter.height = 0.25),
             colour="black", pch=21) +
  scale_radius(range = c(2,10), 
               breaks = c(min(metabolic_env$n),
                          round(mean(metabolic_env$n)),
                          max(metabolic_env$n))) +
  scale_fill_manual(values = MoreColours(length(unique(metabolic_env$Metabolism)))) +
  guides(fill=guide_legend(title="Metabolism"),
         size=guide_legend(title="Count")) +
  ylab("Family") +
  xlab("Ecosystem") +
  theme(panel.background = element_blank(),
        panel.grid.major.y = element_line(colour = "#bdbdbd", linetype = "dotted"),
        panel.grid.major.x = element_line(colour = "#bdbdbd", linetype = "dotted"),
        panel.grid.minor = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1))
```

I finally wanted to know what the distribution of alkane metabolism was for each environment using the McrA phylogenetic signature. All Ecosystems that contribute less that 1/10th of a percentage to the overall total number of McrA classifications have been omitted from the figure.

```{r}
metabolic_env <- mcra_dat %>%
  filter(!Ecosystem.Subtype %in% c("Unclassified", "Unavailable")) %>% 
  group_by(Metabolism, Ecosystem.Subtype) %>% 
  count() %>% 
  ungroup() %>% 
  mutate(Percentage = (100*n)/sum(n)) %>% 
  filter(Percentage >= 0.1) %>% 
  mutate(Ecosystem.Subtype = reorder(Ecosystem.Subtype, Percentage, FUN = sum))

ggplot(metabolic_env, aes(x=Ecosystem.Subtype, y=Percentage, fill=Metabolism)) +
  geom_bar(colour="black", stat = "identity") +
  scale_fill_manual(values = MoreColours(length(unique(metabolic_env$Metabolism)), "RdYlBu")) +
  guides(fill=guide_legend(title="Metabolism"),
         size=guide_legend(title="Count")) +
  ylab("Percentage of Total") +
  xlab("Ecosystem") +
  theme(panel.background = element_blank(),
        panel.grid.major.y = element_line(colour = "#bdbdbd", linetype = "dotted"),
        panel.grid.major.x = element_line(colour = "#bdbdbd", linetype = "dotted"),
        panel.grid.minor = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1))


pred_rank_df <- mcra_dat %>% 
  filter(!Ecosystem.Subtype %in% c("Unclassified", "Unavailable")) %>%
  filter(Resolved != "Root") %>% 
  group_by(Ecosystem.Subtype, Resolved) %>% 
  count() %>% 
  ungroup() %>% 
  mutate(Percentage = (100*n)/sum(n)) %>% 
  filter(Percentage >= 0.1) %>% 
  mutate(Ecosystem.Subtype = reorder(Ecosystem.Subtype, Percentage, FUN = sum)) %>% 
  merge(rank_depth_map, by.x="Resolved", by.y="Rank") %>% 
  mutate(Resolved = reorder(Resolved, Depth))

ggplot(pred_rank_df, aes(x=Ecosystem.Subtype, y=Percentage, fill=Resolved)) +
  geom_bar(colour="black", stat = "identity") +
  scale_fill_manual(values = MoreColours(length(unique(pred_rank_df$Resolved)), "PuOr")) +
  guides(fill=guide_legend(title="Rank Resolved"),
         size=guide_legend(title="Count")) +
  scale_y_reverse() +
  ylab("Percentage of Total") +
  xlab("Ecosystem") +
  theme(panel.background = element_blank(),
        panel.grid.major.y = element_line(colour = "#bdbdbd", linetype = "dotted"),
        panel.grid.major.x = element_line(colour = "#bdbdbd", linetype = "dotted"),
        panel.grid.minor = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1))
```

```{r n_metag, include=FALSE}
wetland_mg <- mcra_dat %>% filter(Ecosystem.Subtype == "Wetlands") %>%
  select(taxon_oid) %>%
  unique() %>% 
  nrow()
vent_mg <- mcra_dat %>% filter(Ecosystem.Subtype == "Hydrothermal vents") %>%
  select(taxon_oid) %>%
  unique() %>% 
  nrow()
```

Firstly, just the wetland and hydrothermal vent metagenomes (`r wetland_mg` and `r vent_mg`, respectively) contribute to 50% of all known McrA sequences. These wetlands are show a predominant hydrogenotrophic signature but all have a strong aceticlastic contingent. Hydrothermal vent communities are mostly alkanotrophic, though specific alkanes currently cannot be discerned phylogenetically while lacking experimentally-confirmed activity annotations. However, the ANME that occupy the methanotrophic clades on the phylogeny are the most significant constituent of these communities relative to Archaea possessing "divergent"-McrA sequences that may facilitate oxidation of higher alkanes.

Further analyses looking into endemism (for example, what environments are the alkanotrophs inhabiting and are these suggestive of a specific niche?) and active-site amino acid profiles are on the to-do list.